### PR TITLE
Splitting Internal and External Enclave Methods

### PR DESCRIPTION
**Pull request by otherview**
### Why this change is needed

Some cleanup for https://github.com/obscuronet/obscuro-internal/issues/1524
This splits the Enclave methods into Internal and External methods.
The external methods are for user/api comsumption.

### What changes were made as part of this PR

Please provide a high level list of the changes made

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/obscuronet/obscuro-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [ ] PR checks reviewed and performed

### CODE REVIEW

File: go/host/rpc/enclaveclient/enclave_client_external.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+		logger:      logger,
+	}
+}
+
+func (c *EnclaveExternalClient) SubmitTx(tx common.EncryptedTx) responses.RawTx {
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), c.config.EnclaveRPCTimeout)
+	defer cancel()
+
+	response, err := c.protoClient.SubmitTx(timeoutCtx, &generated.SubmitTxRequest{EncryptedTx: tx})
+	if err != nil {
```
Comment (BedrockSquirrel) [inline go/host/rpc/enclaveclient/enclave_client_external.go:46]: These 4 lines are duplicated on every call, might be good to wrap them in a standardised method.
```
	if err != nil {
		return responses.AsPlaintextError(err)
	}
	return *responses.ToEnclaveResponse(response.EncodedEnclaveResponse)
```

**Review by BedrockSquirrel — COMMENTED** [review]
Don't fully understand the reasoning behind the split so hard to comment on the naming.

Is the point just to have the methods better organised by splitting the host-driven vs rpc-driven requests or will there be benefits to the implementation from the split too?

File: go/host/rpc/enclaveclient/enclave_client_external.go (hunk 1 above)
Comment (otherview) [reply go/host/rpc/enclaveclient/enclave_client_external.go:46]: Yeap, saw that too, will change that once I introduce the SystemError err.

**Comment by otherview** [comment]
> Is the point just to have the methods better organised by splitting the host-driven vs rpc-driven requests or will there be benefits to the implementation from the split too?

Yeah you got it. Organization and clear splitting of rpc/host methods for this PR.
Splitting the implementation will future help to have wrappers for user responses and system errors, and also helps with making the distinction of allowing host methods to return simple known types while the user responses are typically returned in the JSONRPC compliant structure.

**Comment by otherview** [comment]
> I can't figure out how this will help with the errors.
> 
> I'm ok with the 2 interfaces: `EnclaveUserMethods` and `EnclaveSystemMethods`. But I think we should leave the implementation as it is for now in a single file

Good point, will change to `EnclaveUserMethods` and `EnclaveSystemMethods`.

As for the implementations, is very hard to understand what patterns should be applied in the implementations of `EnclaveUserMethods` vs `EnclaveSystemMethods` if they live in the same file. That's why I split them. 

This is just a stepping stone, not necessarily the final form. I'd prefer to keep the two implementations separate as for sure there will be different patterns in them. One example is that the `enclave_client_external.go` can now have unit tests for the `EnclaveResponse`. And another example is the 4 repeated lines that Matt caught on in https://github.com/obscuronet/go-obscuro/pull/1236#discussion_r1173484190. 

So I think that splitting the implementation is worth doing, but if that doesn't convince you let me know and I'll put them in the same file.

**Comment by tudor-malene** [comment]
> As for the implementations, is very hard to understand what patterns should be applied in the implementations of `EnclaveUserMethods` vs `EnclaveSystemMethods` if they live in the same file. That's why I split them.

You can look at the interface the method implements.
I'd rather not do the splitting. I think this is not a good enough reason for such a change.

File: go/host/container/host_container.go [hunk 2]
```diff
@@ -108,7 +108,7 @@ func NewHostContainerFromConfig(parsedConfig *config.HostInputConfig) *HostConta
 	cfg.ID = ethWallet.Address()
 
 	fmt.Println("Connecting to the enclave...")
-	enclaveClient := enclaverpc.NewClient(cfg, logger)
+	enclaveClient := enclaveclient.NewEnclaveRPCClient(cfg, logger)
```
Comment (tudor-malene) [inline go/host/container/host_container.go:111]: I personally preferred the previous naming. Feels quite redundant now, and doesn't give out more information.

File: go/host/rpc/enclaveclient/enclave_client.go [hunk 3]
```diff
@@ -8,36 +8,26 @@ import (
...
-// Client implements enclave.Enclave and should be used by the host when communicating with the enclave via RPC.
-type Client struct {
-	protoClient generated.EnclaveProtoClient
-	connection  *grpc.ClientConn
-	config      *config.HostConfig
-	logger      gethlog.Logger
-}
-
-func NewClient(config *config.HostConfig, logger gethlog.Logger) *Client {
+func NewEnclaveRPCClient(config *config.HostConfig, logger gethlog.Logger) common.Enclave {
```
Comment (tudor-malene) [inline go/host/rpc/enclaveclient/enclave_client.go:30]: It's hard to follow the reason for these changes.
Stuff has moved around, but is there any change?
Why were the changes/ or moving necessary?

**Review by tudor-malene — COMMENTED** [review]
needs a few more clarifications

**Comment by otherview** [comment]
The idea was to :
1. Clean up enclave methods - split between user and system methods
2. Enclave is now composed of 2 interfaces UserMethods and SystemMethods with their own implementations (as they have different patterns)
3. User methods follow the discussed (EnclaveResponse, SystemErr) pattern
4. System Methods follow another pattern (CustomType, SystemErr)
5. Build tests for User Methods and for System Methods by using the UserMethods implementation and the SystemMethods implementation.

**Comment by otherview** [comment]
Closing in pref of https://github.com/obscuronet/go-obscuro/pull/1237
